cub/block/radix_rank_sort_operations.cuh

```diff
@@ -130,16 +119,518 @@ struct ShiftDigitExtractor : BaseDigitExtractor<KeyT>
 {
     using typename BaseDigitExtractor<KeyT>::UnsignedBits;
 
-    uint32_t bit_start, mask;
+    std::uint32_t bit_start;
+    std::uint32_t mask;
+
     explicit __device__ __forceinline__ ShiftDigitExtractor(
-        uint32_t bit_start = 0, uint32_t num_bits = 0)
-        : bit_start(bit_start), mask((1 << num_bits) - 1)
+        std::uint32_t bit_start = 0, std::uint32_t num_bits = 0)
+      : bit_start(bit_start)
+      , mask((1 << num_bits) - 1)
     { }
 
-    __device__ __forceinline__ uint32_t Digit(UnsignedBits key)
+    __device__ __forceinline__ std::uint32_t Digit(UnsignedBits key) const
+    {
+        return std::uint32_t(this->ProcessFloatMinusZero(key) >> UnsignedBits(bit_start)) & mask;
+    }
+};
+
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS    // Do not document
+namespace detail 
+{
+
+template <bool... Bs>
+struct logic_helper_t;
+
+template <bool>
+struct true_t 
+{
+  static constexpr bool value = true;
+};
+
+template <bool... Bs>
+using all_t =            //
+  ::cuda::std::is_same<  //
+    logic_helper_t<Bs...>, //
+    logic_helper_t<true_t<Bs>::value...>>;
+
+struct fundamental_decomposer_t
+{
+  template <class T>
+  __host__ __device__ T& operator()(T &key) const
+  {
+    return key;
+  }
+};
+
+template<class F, class... Ts, std::size_t... Is>
+__host__ __device__ void
+for_each_member_impl_helper(F f, const ::cuda::std::tuple<Ts&...>& tpl, THRUST_NS_QUALIFIER::index_sequence<Is...>)
+{
+  auto sink = { (f(::cuda::std::get<Is>(tpl)), 0)... };
+  (void)sink;
+}
+
+template<class F, class... Ts>
+__host__ __device__ void
+for_each_member_impl(F f, const ::cuda::std::tuple<Ts&...>& tpl)
+{
+  static_assert(sizeof...(Ts), "Empty aggregates are not supported");
+
+  // Most radix operations are indifferent to the order of operations.
+  // Conversely, the digit extractor traverses fields from the least significant 
+  // to the most significant to imitate bitset printing where higher bits are on 
+  // the left. It also maps to intuition, where something coming first is more 
+  // important. Therefore, we traverse fields on the opposite order.
+  for_each_member_impl_helper(f, tpl, THRUST_NS_QUALIFIER::make_reversed_index_sequence<sizeof...(Ts)>{});
+}
```

Comment:
Compare compilation time performance of fold expressions to see if there's a value in conditionally using them here.

Reply:
Compared with:
```cpp
(f(::cuda::std::get<Is>(tpl)), ...);
```
No difference in SASS, compilation time match current implementation.